internal/gatewayapi/securitypolicy.go

```diff
@@ -828,6 +911,46 @@ func (t *Translator) translateSecurityPolicyForGateway(
 			}
 		}
 	}
+
+	// Pre-create a TCP-only authorization object to avoid re-allocation
+	var tcpAuthorization *ir.Authorization
+	if authorization != nil {
+		authCopy := *authorization
+		tcpAuthorization = &authCopy
+	}
+
+	// Apply to TCP listeners (Authorization only). Support metadata nil fallback by parsing section name from listener name suffix.
+	if tcpAuthorization != nil {
+		for _, tl := range x.TCP {
+			if tl == nil || len(tl.Routes) == 0 {
+				continue
+			}
+			// TCPListener name has same format namespace/gatewayName/listenerName
+			gatewayNameEnd := strings.LastIndex(tl.Name, "/")
+			if gatewayNameEnd <= 0 {
+				continue
+			}
+			gatewayName := tl.Name[0:gatewayNameEnd]
+			if t.MergeGateways && gatewayName != policyTarget {
+				continue
+			}
+			listenerSectionName := ""
+			if tl.Metadata != nil && tl.Metadata.SectionName != "" {
```

Comment:
Can't we trust the tl.Metadata.SectionName like the HTTP listener does? Can we use the same approach for both the TCP and HTTP?